### DISCUSSION TITLE
Changing a foreign key

### DISCUSSION THREAD

**Discussion by francoo98** [sqlalchemy/sqlalchemy#12194] [Usage Questions]
Hello, I have these models:
```python
from typing import List
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column
from sqlalchemy.orm import relationship
from sqlalchemy import String, ForeignKey


class Base(DeclarativeBase):
    pass


class Parent(Base):
    __tablename__ = "Parent"

    id: Mapped[int] = mapped_column(primary_key=True)
    name: Mapped[str] = mapped_column(String(50), unique=True)
    children: Mapped[List["Child"]] = relationship(back_populates="parent")


class Child(Base):
    __tablename__ = "Child"

    parent_name: Mapped[str] = mapped_column(ForeignKey("Parent.name",
                                                        onupdate="CASCADE"),
                                             primary_key=True)
    parent: Mapped["Parent"] = relationship(back_populates="children")

```
And this code:
```python
from sqlalchemy.orm import Session
from sqlalchemy import create_engine
import models


engine = create_engine(("mysql+pymysql://"
                        "root:pass"
                        "@"
                        "127.0.0.1:45000/"
                        "prueba"),
                       echo="debug")

models.Base.metadata.drop_all(engine)
models.Base.metadata.create_all(engine)


with Session(engine) as session:
    parent = models.Parent()
    parent.name = "asd"
    child = models.Child()
    child.parent_name = parent.name
    parent.children.append(child)
    session.merge(parent)
    session.commit()

with Session(engine) as session:
    parent = models.Parent()
    parent.id = 1
    parent.name = "kkk"
    child = models.Child()
    child.parent_name = parent.name
    parent.children.append(child)
    session.merge(parent)
    session.commit()

```
Notice that in the second `with` block, I'm creating an object and setting its id to one instead of reading it from the db. I have to do this this way because, in my app, I'm reading objects from a secondary source other than the db. So, I'm reading data, creating an object, setting all its attributes, and then merging it. This approach fails when I need to rename a parent.

This is also the reason for the re-creation of the children in this block. This is what fails and gives me the following error:

```
AssertionError: Dependency rule on column 'Parent.name' tried to blank-out primary key column 'Child.parent_name' on instance '<Child at 0x771b8927fa70>'
```
Stackstrace
```
2024-12-19 11:54:05,766 INFO sqlalchemy.engine.Engine SELECT `Parent`.id AS `Parent_id`, `Parent`.name AS `Parent_name` 
FROM `Parent` 
WHERE `Parent`.name = %(param_1)s
2024-12-19 11:54:05,766 INFO sqlalchemy.engine.Engine [generated in 0.00031s] {'param_1': 'asd'}
2024-12-19 11:54:05,767 DEBUG sqlalchemy.engine.Engine Col ('Parent_id', 'Parent_name')
2024-12-19 11:54:05,768 INFO sqlalchemy.engine.Engine ROLLBACK
Traceback (most recent call last):
  File "path/main.py", line 34, in <module>
    session.commit()
  File "path/venv/lib/python3.12/site-packages/sqlalchemy/orm/session.py", line 2028, in commit
    trans.commit(_to_root=True)
  File "<string>", line 2, in commit
  File "path/venv/lib/python3.12/site-packages/sqlalchemy/orm/state_changes.py", line 139, in _go
    ret_value = fn(self, *arg, **kw)
                ^^^^^^^^^^^^^^^^^^^^
  File "path/venv/lib/python3.12/site-packages/sqlalchemy/orm/session.py", line 1313, in commit
    self._prepare_impl()
  File "<string>", line 2, in _prepare_impl
  File "path/venv/lib/python3.12/site-packages/sqlalchemy/orm/state_changes.py", line 139, in _go
    ret_value = fn(self, *arg, **kw)
                ^^^^^^^^^^^^^^^^^^^^
  File "path/venv/lib/python3.12/site-packages/sqlalchemy/orm/session.py", line 1288, in _prepare_impl
    self.session.flush()
  File "path/venv/lib/python3.12/site-packages/sqlalchemy/orm/session.py", line 4352, in flush
    self._flush(objects)
  File "path/venv/lib/python3.12/site-packages/sqlalchemy/orm/session.py", line 4487, in _flush
    with util.safe_reraise():
  File "path/venv/lib/python3.12/site-packages/sqlalchemy/util/langhelpers.py", line 146, in __exit__
    raise exc_value.with_traceback(exc_tb)
  File "path/venv/lib/python3.12/site-packages/sqlalchemy/orm/session.py", line 4448, in _flush
    flush_context.execute()
  File "path/venv/lib/python3.12/site-packages/sqlalchemy/orm/unitofwork.py", line 466, in execute
    rec.execute(self)
  File "path/venv/lib/python3.12/site-packages/sqlalchemy/orm/unitofwork.py", line 591, in execute
    self.dependency_processor.process_saves(uow, states)
  File "path/venv/lib/python3.12/site-packages/sqlalchemy/orm/dependency.py", line 604, in process_saves
    self._synchronize(
  File "path/venv/lib/python3.12/site-packages/sqlalchemy/orm/dependency.py", line 625, in _synchronize
    sync.clear(dest, self.mapper, self.prop.synchronize_pairs)
  File "path/venv/lib/python3.12/site-packages/sqlalchemy/orm/sync.py", line 88, in clear
    raise AssertionError(
AssertionError: Dependency rule on column 'Parent.name' tried to blank-out primary key column 'Child.parent_name' on instance '<Child at 0x771b8927fa70>'
```
Is there a way to make this code work? Thanks in advanced.
I'm using sqlalchemy 2.0.36 and mariadb 11.5.2-MariaDB-ubu2404

If you need to run the code, you can use this `docker-compose.yml`:
```yaml
version: '3.1'

services:

  db:
    image: mariadb
    restart: always
    environment:
      MARIADB_ROOT_PASSWORD: pass
    ports:
      - 45000:3306

  adminer:
    image: adminer
    restart: always
    ports:
      - 46000:8080
```

**Comment by zzzeek** [comment]
a couple of things are happening

first note that because you are merging on id=1, which is the autogenerated id of the original Parent, there is only one `Parent` row we are dealing with here.  not sure if that's clear or if that is your intent.

So when you merge that new Parent object that matches the original row, it has a `.children` collection with only a new `Child` in it, not the old one.  That means you are wishing to remove the first `Child` from that collection.  SQLAlchemy follows your instruction and nulls out the `parent_id` column on that first child which fails because the column is not nullable.  you would need to decide what behavior you want here:

1. **i dont want to remove the Child** -  you likely dont want to merge a new `Parent` with empty `.children` collection, you'd merge first, load `.children`, then append new `Child`, or, just dont touch `parent.children` in the second run, it's not needed to do the INSERT you are looking for, leave it unset
2. **I want to remove the Child and have it be deleted** - look into `cascade='all, delete-orphan'`, see https://docs.sqlalchemy.org/en/20/orm/cascades.html#delete
3. **I want the Child to have a null parent_name** - change the column to be nullable

There's another thing going on here as well which is that you are changing the name of that first `Parent`, which is a unique col with FK cols referring to it.   To do this in the database without an FK violation you need to set that FK to be [on update cascade](https://docs.sqlalchemy.org/en/20/orm/relationship_persistence.html#passive-updates) and likely also use `passive_updates` on that relationship .  

overall if there's a way you can load the Parent you want to update first, then manually populate without using merge things might be simpler here, plus the mutable FK constraint thing is also not ideal if you have a choice on that

**Reply by francoo98** [reply]
Hi, thank you for taking your time to answer 🙏
I was expecting SQLAlchemy to notice that the `Child` in the code is the one in the db, but I now notice there is nothing suggesting that.
I want to go with `all, delete-orphan`, so the old `Child` gets deleted and then the new one (the one in the session block) gets added with the new name.
But, this gives me the following error:
```
sqlalchemy.exc.IntegrityError: (pymysql.err.IntegrityError) (1062, "Duplicate entry 'kkk' for key 'PRIMARY'")
[SQL: INSERT INTO `Child` (parent_name) VALUES (%(parent_name)s)]
[parameters: {'parent_name': 'kkk'}]
```
I'm afraid the old `Child` is not orphaned right? Its name is changed.

> plus the mutable FK constraint thing is also not ideal if you have a choice on that

I think I'll have to go with that.

**Reply by zzzeek** [reply]
OK you are coming across what's essentially the only shortcoming in the unit of work's design which is known as the "DELETE before INSERT" problem which is at #2501

your mapping and workflow here is literally a perfect storm of factors to cause this issue, including that not only is the parent.name a unique constraint but not the primary key (which the ORM would be able to handle), but that the Child.parent_name **is** a primary key.    The ORM can easily miss that Child row in many ways since this is a very far edge case for the ORM.

In order to get the unit of work to do this, you would need to run these operations more individually, in some way do the collection maintenance first, then do the name change, and I would make sure your code is thoroughly tested - after trying a few different ways and seeing warnings or failures of various forms, I was able to get this version to work

```py
with Session(engine) as session:
    parent = Parent(id=1, children=[])
    parent = session.merge(parent)
    session.flush()
    print(session.execute(text("select * from Child")).all())

    parent.name = "kkk"
    child = Child()
    parent.children.append(child)
    session.commit()

    print(session.execute(text("select * from Child")).all())
```

however, this all seems like overkill.   you have the foreign key cascades in place, you really just need to run a single UPDATE.  doing this works a lot more simply:


```py
from sqlalchemy import update

with Session(engine) as session:
    session.execute(
        update(Parent).values(name="kkk").where(Parent.id==1)
    )

    print(session.execute(text("select * from Child")).all())
```

since that UPDATE is really all you need here.   you can also do this using [bulk ORM update format](https://docs.sqlalchemy.org/en/20/orm/queryguide/dml.html#orm-queryguide-bulk-update) where you can just pass dictionaries

```py
with Session(engine) as session:
    session.execute(update(Parent), [{"name": "kkk", "id": 1}])

    print(session.execute(text("select * from Child")).all())
```
